### ISSUE TITLE
ROI tool: Fix fast mouse movements with large brush

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#203]
Fast mouse movements can introduce gaps in the ROI drawn. This was fixed for the case where the size of the brush is minimal, by essentially ray-tracing through the voxel grid. This could be extended to work for any brush size, by combining a sphere draw at each end of the ray segment with a search for voxels whose minimum tangential distance from the ray segment are less than the brush width.

**Comment by rtabbara** [comment]
@Lestropie @jdtournier Had a go at this --- not sure if it was exactly what was in mind. Per mouse event, we interpolate between the previous and current positions over a fixed time step, iteratively drawing a circle centered around the interpolated point.

It's simple but seems to work quite well over a range of different widths/resolutions. If we wanted to be even fancier, we could make the time step a function of the brush diameter and point distance to avoid overdraw, but to be honest I don't think the current performance is causing any kind of bottleneck.

And yes, if you let the mouse *really* rip, you might observe some gaps, however I don't think (hopefully) that would be a typical use case (?)

**Comment by Lestropie** [comment]
Not really a 'clean' solution; may still have gaps for very fast movements, and is doing more processing than necessary for small movements.

It should be possible to combine a sphere draw at each endpoint with a cylinder draw between them:
* Identify voxels along the ray between the two points, and use these voxels as the initialisation for an outward voxel search
* Voxels are set (and their neighbours added to the search space) if they:
  - Lie between the two endpoints along the length of the cylinder:
`((p-e1).dot(e2-e1) > 0 && (p-e1).dot(e2-e1) < (e2-e1).norm()`
, and
  - Tangential distance to ray is less than the cylinder radius:
`((p-e1) - (p-e1).dot((e2-e1).normalise())).norm() < rad`

(may be faster ways to calculate those... that's just how it works in my head)

**Comment by draffelt** [comment]
Rob, I'm pretty sure doing all those calculations would be more expensive
than overdrawing a few times on the GPU. Like Rami said, you need to be
moving the mouse at a ridiculous speed to see any gaps, something that no
user would do in practice.

**Comment by Lestropie** [comment]
Sphere overdraw is occurring on the CPU.

Should actually be _less_ expensive my way; each voxel within the plausible region would be processed only once, albeit with a more expensive operator, whereas with overdraw each could be tested up to 10 times.

I can implement it myself down the track if need be; I just don't see the point in using a 'hack' solution when a precise solution is available and doesn't take a whole lot more work.

**Comment by jdtournier** [comment]
Like @Lestropie says, no ROI editing is happening on the GPU. I tried very hard to have everything happen on the GPU, but there is no way to do this without OpenGL4 features... and we've already had that discussion around issue #91. 

I agree with @Lestropie that it's not the prettiest solution, but if it works and isn't a performance hog, then let's leave it in for now. We can always modify later when we've dealt with all the other issues... :wink: I mean, let's not forget that this is really very much a non-essential bit of polish - nice to have, but no-one is going to claim their money back because of this... That said, the approach suggested by @Lestropie is how I'd approach it, and it should be relatively straightforward. 

The other quick fix we can do in the meantime is simply to be a bit smarter about the increment in that loop. Currently it simply draws 10 circles at regular intervals between the end-points. It should be trivial to calculate the distance between the end-points in voxel units, round to nearest, and use the reciprocal of that as the increment...?

**Comment by rtabbara** [comment]
No problem. For the record, if there's a more robust and efficient approach to doing this then I'm all for it being implemented.
